### PR TITLE
[RISCV] Remove CalleeSavedInfo for Zcmp/save-restore-libcalls registers

### LINKED ISSUE #79491: [RISCV] Post-PEI MIR testing is broken with Zcmp/libcalls
MIR printing is broken when trying to serialize the output after PrologEpilogInserter (`-run-pass=prologepilog` or `-stop-after`).

`llc -mtriple=riscv32 -mattr=+zcmp -x mir -run-pass=prologepilog -verify-machineinstrs -o - `[zcmp-cm-push-pop.mir](https://github.com/llvm/llvm-project/files/14056737/zcmp-cm-push-pop.mir.txt):

```
Assertion failed: (unsigned(ObjectIdx+NumFixedObjects) < Objects.size() && "Invalid Object Idx!"), function isDeadObjectIndex, file MachineFrameInfo.h, line 747.
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
```

Zcmp/save-restore-libcalls have pre-defined frame indices for the saved registers but they're never allocated in `MachineFrameInfo`. When MIRPrinting occurs, the index found in the entries in `MachineFrameInfo::CalleeSavedInfos` is not to be found in `MachineFrameInfo::Objects`, and `isDeadObjectIndex` asserts.

We should create proper (fixed?) slots in MFI in the future to avoid inconsistencies there, since we don't really care about the frame index to be fixed, but the location on the stack.

### PR DESCRIPTION
**Pull request by francisvm**
Registers that are pushed/popped by Zcmp or libcalls have pre-defined frame indices that are never allocated in MachineFrameInfo. They're being used throughout PEI, but the rest of codegen doesn't work that way and expects each frame index to be a valid index in MFI.

This patch keeps it local to PEI and removes them from the CalleeSavedInfo list at the end of the pass.

Before this pass, any MIR testing post-PEI is broken and asserts (see issue #79491).

### CODE REVIEW

**Comment by wangpc-pp** [comment]
Thanks! I have a question:
Does ARM have the same issue as it has PUSH/POP? How does it handle these CSIs?

**Comment by francisvm** [comment]
> Thanks! I have a question: Does ARM have the same issue as it has PUSH/POP? How does it handle these CSIs?

No, it doesn't have this issue. The issue here is that in MFI, there are no real stack objects created for these CSR spills, which is what a frame index is supposed to point to. In the other backends, they get real stack objects with frame indices that are not pre-defined, which is what we should probably do here too. I did try to fix it properly but there are too many assumptions made around the current mode that makes it a little harder to fix for me.

File: llvm/lib/Target/RISCV/RISCVFrameLowering.cpp [hunk 1]
```diff
@@ -1252,6 +1252,30 @@ void RISCVFrameLowering::processFunctionBeforeFrameFinalized(
...
 }
 
+void RISCVFrameLowering::processFunctionBeforeFrameIndicesReplaced(
+    MachineFunction &MF, RegScavenger *RS) const {
+  // Remove CalleeSavedInfo for registers saved by Zcmp or save/restore
+  // libcalls.
+  MachineFrameInfo &MFI = MF.getFrameInfo();
+  const TargetRegisterInfo *TRI = MF.getSubtarget().getRegisterInfo();
+  const auto *RVFI = MF.getInfo<RISCVMachineFunctionInfo>();
+  if (!RVFI->isPushable(MF) && !RVFI->useSaveRestoreLibCalls(MF))
```
Comment (wangpc-pp) [inline llvm/lib/Target/RISCV/RISCVFrameLowering.cpp:1262]: We need test coverage for `useSaveRestoreLibCalls`.

**Review by wangpc-pp — APPROVED** [review]
The code looks fine, but I am not familiar with this part, so please wait for another approval.

File: llvm/lib/Target/RISCV/RISCVFrameLowering.cpp (hunk 1 above)
Comment (francisvm) [reply llvm/lib/Target/RISCV/RISCVFrameLowering.cpp:1262]: Added test with `-mattr=+save-restore`

File: llvm/lib/Target/RISCV/RISCVFrameLowering.cpp [hunk 2]
```diff
@@ -1252,6 +1252,30 @@ void RISCVFrameLowering::processFunctionBeforeFrameFinalized(
...
+    return;
+  const std::vector<CalleeSavedInfo> &CSIs = MFI.getCalleeSavedInfo();
+  std::vector<CalleeSavedInfo> NewCSIs;
+  for (const auto &CSI : CSIs) {
+    // Skip CSRs that have fake a frame index.
+    [[maybe_unused]] int ReservedFI = 0;
```
[the remark this answers is not in the document]
Comment (francisvm) [reply llvm/lib/Target/RISCV/RISCVFrameLowering.cpp]: Removed, thanks!

**Comment by francisvm** [comment]
Last push should fix the test. Let me know if anyone has any more concerns, I'd like to put up other patches that need MIR testing.

**Comment by francisvm** [comment]
ping?

**Review by yetingk — APPROVED** [review]
LGTM.

**Comment by topperc** [comment]
Could this patch affect Inter Procedural Register Allocation? I'm seeing a regression and I'm guess this could be related.

**Comment by francisvm** [comment]
> Could this patch affect Inter Procedural Register Allocation? I'm seeing a regression and I'm guess this could be related.

hmm, PEI runs post-ra, right? I don't know how IPRA works and what is the state of each function when it runs, but this should only be apparent post-PEI.

It does seem like `LiveRegUnits` uses `MFI.getCalleeSavedInfo` for pristine registers, so maybe there is something related there?

```
llvm/lib/CodeGen/LiveRegUnits.cpp
102:    const auto &CSI = MFI.getCalleeSavedInfo();
122:    for (const CalleeSavedInfo &Info : MFI.getCalleeSavedInfo())
133:  for (const CalleeSavedInfo &Info : MFI.getCalleeSavedInfo())
```

**Comment by topperc** [comment]
> > Could this patch affect Inter Procedural Register Allocation? I'm seeing a regression and I'm guess this could be related.
> 
> hmm, PEI runs post-ra, right? I don't know how IPRA works and what is the state of each function when it runs, but this should only be apparent post-PEI.
> 
> It does seem like `LiveRegUnits` uses `MFI.getCalleeSavedInfo` for pristine registers, so maybe there is something related there?
> 
> ```
> llvm/lib/CodeGen/LiveRegUnits.cpp
> 102:    const auto &CSI = MFI.getCalleeSavedInfo();
> 122:    for (const CalleeSavedInfo &Info : MFI.getCalleeSavedInfo())
> 133:  for (const CalleeSavedInfo &Info : MFI.getCalleeSavedInfo())
> ```

IPRA has a pass that collects which registers a function uses. That is used to update the regmask for call sites in its callers. I think that collection pass runs after PEI.

**Comment by topperc** [comment]
I'm going to attempt to register the slots with MFI.

**Comment by francisvm** [comment]
> I'm going to attempt to register the slots with MFI.

that would be great, feel free to add me as a reviewer

**Pull request merged**